### ISSUE TITLE
alias should all check for `command foo` when generating `--wraps` to prevent recursion

### ISSUE THREAD

**Issue by mattmc3** [fish-shell/fish-shell#9493]
I got bit by the "error: completion reached maximum recursion depth, possible cycle?" alias issue today. I couldn't find this exact issue reported here, but others have definitely [experienced it](https://unix.stackexchange.com/questions/675018/fish-maximum-recursion-depth-reached-with-tab-complete). Essentially, if you create an alias to override a built-in using `command foo`, the `--wraps` statement gets built in such a way as to confuse completion and causes a `max recursion` issue.

I got it by creating this alias: `alias grep "command grep --color=auto --exclude-dir={.git,.hg,.svn}"`.

The alias.fish function tries to detect recursion, but doesn't specifically look for the `command` keyword.

https://github.com/fish-shell/fish-shell/blob/d50e9ffff35a12b542f41e1bdf1ad5357fba7ead/share/functions/alias.fish#L61-L68

A possible simple solution is to add an additional check: `test $first_two_words != "command $name"`

This is similar to, but not the same as #7389 and #6899.

The workaround is to change the `--wraps` manually after running `funcsave foo`.

**Comment by faho** [comment]
First of all: Write a function.

All issues with alias go away when you just write a function and I am *this close* to just removing it entirely and getting yelled at a bunch.

Secondly: the big problem here is that we need to figure out the "actual command", and enumerating decorators seems to ve fragile and a waste of time.

Tbh I'm starting to think this needs support in commandline.

**Comment by mattmc3** [comment]
> First of all: Write a function.

Agree. `alias`, as implemented, is flawed and slow and seems just there to make it easier to migrate from other shells. I'd be happy to see it removed, or just simplified to a `funcsave` wrapper. BUT - since it's here and ships with fish and behaves badly, I made this issue as much for SEO as anything else to help future travelers who get bit by this.

> Tbh I'm starting to think this needs support in commandline - commandline --current-process --command, which will give either just "git" for "true; and command git -C foo status", or "and command git" (so the "actual" command would always be the last part, but you can get the decorators if you need them).

You would know better than me. In my simplistic view, it just seemed like `--wraps` was the only broken part here.

**Comment by Quantenzitrone** [comment]
Can someone explain, why `function mycommand --wraps="command mycommand"` recurses infinitely?
To me, it would make more sense if `function mycommand --wraps="mycommand"` recursed, because wouldn't `command` be the one stopping the recursion by ignoring the defined functions and builtins and executing (or in this case wrapping) the program instead.

**Comment by faho** [comment]
>it would make more sense if function mycommand --wraps="mycommand" recursed

It would, except the completion wrapping code recognizes that case - it'll see that it's exactly the same thing and stop recursing.

>Can someone explain, why function mycommand --wraps="command mycommand" recurses infinitely?

It doesn't recurse infinitely, it'll hit the maximum limit. You don't get a fish that hangs, you get an error message. 

And it does that because the wrapping code looks at "command mycommand" and tries to complete it like "command", and the "command" completion tells it to complete like its subcommand. There's one more step and the loop detection doesn't catch it.

In general any loop detection will unfortunately always be imperfect - you'll always be able to construct a case where completion recurses in a way that the loop detection doesn't see as a loop.

>wouldn't command be the one stopping the recursion by ignoring the defined functions and builtins and executing (or in this case wrapping) the program instead.

These are separate things - "command" just says to complete like its first argument, and the completions don't care about whether it's a function or a builtin, and it doesn't really matter either - you can do `complete --command grep --wraps "command grep"` without any function involved at all.

It's just that that's obviously a useless thing to do and so people generally don't, while `alias`, that lovely function, isn't smart enough to figure it out in this case.

And so the nicer solution is to stop it *there*, with a feature that would also be useful for other things (like abbreviations that only work for arguments to certain commands)

**Comment by Quantenzitrone** [comment]
I think it would make sense for the command completion to behave like the execution for `command` and `builtin`.
So `command` should tell the completion to ignore completions for builtins or functions, and search for a program instead.
I rather explicitly tell my alias/function to use the builtin or the executable program instead of some weird loop detection mechanism doing that for me.
While i agree that aliases could be removed in favor of encouraging users to use functions instead, having `alias ls="ls --color=tty"` working without any recursion problems, but `alias ls="command ls --color=tty"` running into max recursion depth is just really trippy (unintuitive).

**Comment by faho** [comment]
>I think it would make sense for the command completion to behave like the execution for command and builtin.

The completion *script* does not *care* what sort of thing it is completing. It simply does not operate on that level. It doesn't "search".

If you do `complete --command grep --long-option help` that will complete "--help" when told to complete for "grep" - whether that is a function, a command or a builtin. It doesn't need to know!

And changing this would not only be a big complication, it also would not help with this issue. See above.

(there are cases where you would want to have "the exact thing we're completing right now" - like when you complete `./foo` and that's a different version or thing from `/usr/bin/foo`, but that doesn't matter for this issue - #3117)

>having alias ls="ls --color=tty" working without any recursion problems, but alias ls="command ls --color=tty" running into max recursion depth is just really trippy (unintuitive).

Of course, and that is why this issue is still open.

**Comment by adhdtiger** [comment]
So I found a weirdness.  I was getting the "error: completion reached maximum recursion depth, possible cycle?" error with vi <tab> completion.  Because of muscle memory I have vi functioned to a different editor and I was getting this error seeming no matter what I tried.  Well, turns out if --description actually has your alias command that you ported over it tries to set an actual alias.

I had half a dozen functions where the pattern was 
'''
function cmd --wraps=desired_cmd --description 'alias cmd=desired_cmd'
    desired_cmd $argv
end
'''
and the result was I could run 
`alias`
on the command line and it would list any description from my functions that had the 'alias cmd=other_cmd' pattern. 

Once I cleaned the word alias from all descriptions, my tab completion on my 'vi' function started working properly without the recursion error and 'alias' stopped printing out results.

**Comment by faho** [comment]
> Well, turns out if --description actually has your alias command that you ported over it tries to set an actual alias.

It does not. Fish does not really *have* aliases. `alias` is just a cheesy function that creates functions.

And to figure out what functions "are" aliases for when you run `alias` to list them, it checks the description. But that does not alter the functions in any way.

**Comment by adhdtiger** [comment]
Just the behavior I witnessed. Until I removed "alias" from the descriptions, my `function vi` would return the recursion error.   Once just the word 'alias' was removed from my function's description, the recursion errors ceased. and it was the only change I made.

I had used the fish `alias -s` command as a way to quickly edit my zsh aliases file and get them entered as functions for fish. Most of them I had gone back and removed much of the generated description to make the code more readable.  vi was one that I hadn't because the alias command was really short.

Good to know about the reporting of the description when executing alias without any arguments.
